feat: put method in http postprocessor

Batrov opened the pull request:
Add PUT functionality to http postprocessor

prakharmathur82 commented on dagger-core/src/main/java/io/odpf/dagger/core/processors/external/http/HttpSourceConfig.java:
```diff
@@ -90,6 +93,16 @@ public String getEndpoint() {
+     *
+     * @return the endpointVariables
+     */
+    public String getEndpointVariables() {
```
we should use this for Get Request as well, but we will need to make it backwards compatible

Batrov replied:
updated

Meghajit replied:
Also, I have a concern that this might break the existing running daggers when deployed. 

`Reason`:  As of now, the config `PROCESSOR_POSTPROCESSOR_CONFIG` looks something like below ( example)

```
{
  "external_source": {
	"http": [
    {
      "endpoint": "http://localhost:8080/match",
      "verb": "post",
      "request_pattern": "{\"locations\":%s,\"hints\”:{\”type\”:\”car\”,\”isold\”:false,\”isnew\”:true,\”percentage\”:50}}",
      "request_variables": “my_location”,
      "stream_timeout": "100000",
      "connect_timeout": "10000",
      "fail_on_errors": "false",
      "capacity": "30",
      "headers": {
        "content-type": "application/json"
      },
      "output_mapping": {
        “response_data”: {
          "path": "$"
        },
        “my_stage”: {
          "path": “$.stage”
        }
      }
    }
  ]
  }
}
```

This JSON config value is then deserialized into `PostProcessorConfig`  [here](https://github.com/odpf/dagger/blob/e46f634ca87de896b701ecd0acf1832b4b73ebdd/dagger-core/src/main/java/io/odpf/dagger/core/processors/PostProcessorConfig.java#L48). AFAIK, for existing daggers `endpointVariables` will then be deserialized into `null`. This may cause a `NullPointerException` when you are checking like this :

```
   if (httpSourceConfig.getEndpointVariables().length() > 0) {
            endpoint = String.format(httpSourceConfig.getEndpoint(), httpSourceConfig.getEndpointVariables());
        }
```

Batrov replied:
> Any reason why endpointVariables is a String and not an Object[] like requestVariables ?

Can you point me where requestVariables is an Object[]? Since based on [this](https://github.com/odpf/dagger/blob/main/dagger-core/src/main/java/io/odpf/dagger/core/processors/external/http/HttpSourceConfig.java#L22), it's also a string

> How will this look like ? Can you give an example value ?

I've added test case [here](https://github.com/odpf/dagger/pull/217/commits/1334756642f2a3246070b0c8c90aa74c70825f5c#diff-c7a2518814bf447e7ab2b82bce29b92285113176b5acbb3163d8e7ea3a9d7c38R295) on how will it looks like


> This may cause a NullPointerException

true, updated to (!StringUtil.isNullOrEmpty(httpSourceConfig.getEndpointVariables()))

Meghajit replied:
> Can you point me where requestVariables is an Object[]? Since based on [this](https://github.com/odpf/dagger/blob/main/dagger-core/src/main/java/io/odpf/dagger/core/processors/external/http/HttpSourceConfig.java#L22), it's also a string

My bad. I meant `dynamicHeaderVariablesValues` and `requestVariablesValues`. 

If you check [here](https://github.com/odpf/dagger/blob/main/dagger-core/src/main/java/io/odpf/dagger/core/processors/external/http/HttpAsyncConnector.java#L97), we parse the string input such as `ExternalPostProcessorVariableType.HEADER_VARIABLES` and `ExternalPostProcessorVariableType.REQUEST_VARIABLES` into an Object[].  This is because user can specify multiple variables in the string like `customer_name, order_number`.

It depends on whether we should allow user to configure multiple variables in the endpoint also. I think it should be allowed :-) 


> true, updated to (!StringUtil.isNullOrEmpty(httpSourceConfig.getEndpointVariables()))
If we make it as an Object[ ], this check won't be needed anymore.

Batrov replied:
updated the PR @Meghajit

Meghajit replied:
Can you add a unit test for endpoint with multiple variables as well ?

Batrov replied:
https://github.com/odpf/dagger/pull/217/files#diff-068c730b8c7790e141bb473ed08ae9ef6d60558d51288be41c10a5dcc9754321R67 added one here

Meghajit replied:
Thanks. Looks fine.

Meghajit commented:
We should not commit this file `dagger-common/src/generated-sources/descriptors/dagger-descriptors.bin`

Batrov commented:
> We should not commit this file `dagger-common/src/generated-sources/descriptors/dagger-descriptors.bin` It's an autogenerated one and only used for quickstart. I'll create an issue to add this in `.gitignore`

updated and added in gitignore

Meghajit commented:
We will need to add this new feature in the documentation as well. 
Can you also add/edit the necessary sections [here](https://odpf.github.io/dagger/docs/advance/post_processor#http) ?

Meghajit commented on dagger-core/src/main/java/io/odpf/dagger/core/processors/external/http/request/HttpPostRequestHandler.java:
```diff
@@ -39,7 +42,7 @@ public BoundRequestBuilder create() {
         String endpoint = httpSourceConfig.getEndpoint();
 
         if (!StringUtil.isNullOrEmpty(httpSourceConfig.getEndpointVariables())) {
```
You need not do this validation check anymore as `endpointVariablesValues` cannot be null but can be empty.
You can just do 
`String endpoint = String.format(httpSourceConfig.getEndpoint(), endpointVariablesValues)` 

at line 42.

Meghajit commented on dagger-core/src/main/java/io/odpf/dagger/core/processors/external/http/request/HttpPutRequestHandler.java:
```diff
@@ -39,7 +42,7 @@ public BoundRequestBuilder create() {
         String endpoint = httpSourceConfig.getEndpoint();
 
         if (!StringUtil.isNullOrEmpty(httpSourceConfig.getEndpointVariables())) {
```
You need not do this validation check anymore as endpointVariablesValues cannot be null but can be empty.
You can just do
`String endpoint = String.format(httpSourceConfig.getEndpoint(), endpointVariablesValues)`

at line 42.

Batrov replied:
https://github.com/odpf/dagger/pull/217/files#diff-068c730b8c7790e141bb473ed08ae9ef6d60558d51288be41c10a5dcc9754321R67 I added one here @Meghajit

Meghajit commented on dagger-core/src/main/java/io/odpf/dagger/core/processors/external/http/request/HttpGetRequestHandler.java:
```diff
@@ -40,7 +43,7 @@ public BoundRequestBuilder create() {
         String endpoint = httpSourceConfig.getEndpoint();
 
         if (!StringUtil.isNullOrEmpty(httpSourceConfig.getEndpointVariables())) {
```
You need not do this validation check anymore as endpointVariablesValues cannot be null but can be empty.
You can just do
String endpoint = String.format(httpSourceConfig.getEndpoint(), endpointVariablesValues)

at line 43.

Batrov replied:
removed that validation
